### PR TITLE
Client code refactoring

### PR DESCRIPTION
**Pull request by bsekachev**
### Motivation and context
* I converted (cvat-data, cvat-core, cvat-canvas, cvat-canvas3d) to ESM. In fact they already were ESM, but still considered as CommonJS modules without explicit ``{ "type": "module" }`` in package.json. 
* CommonJS files renamed from .js to .cjs within ESM modules (it is a part of specification)
* I added ``cvat-core`` global types, now ``getCore()`` returns typed object 
* Updated browserlist config, Firefox < 102 is not used on the market according to browserlist info. I would like also update spec for Chrome for more modern version, but these old version, like 63, 64, 65 are still used by users around the world :(
* Updated many eslint packages, minor code refactoring related to these changes, updated linting rules
* And actually many other packages also were updated
* A little bit refactored jest tests, I tried to minimize using ``window`` there. I hope soon we will be able to get rid of ``jsdom`` dependency. Right now 3rdparty packages from ``cvat-data`` don't allow to do that
* Removed extra code from ``cvat-data`` module.
* Supported typescript workers with Webpack 5, removed worker loader
* Removed events.subscribe, events.unsubscribe (we did not even have implementation for these cvat-core methods) 

### Checklist

- [x] I submit my changes into the `develop` branch
- [ ] I have created a changelog fragment 
- [ ] I have updated the documentation accordingly
- [ ] I have added tests to cover my changes
- [ ] I have linked related issues (see [GitHub docs](
  https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword))
- [ ] I have increased versions of npm packages if it is necessary
  ([cvat-canvas](https://github.com/opencv/cvat/tree/develop/cvat-canvas#versioning),
  [cvat-core](https://github.com/opencv/cvat/tree/develop/cvat-core#versioning),
  [cvat-data](https://github.com/opencv/cvat/tree/develop/cvat-data#versioning) and
  [cvat-ui](https://github.com/opencv/cvat/tree/develop/cvat-ui#versioning))

### License

- [x] I submit _my code changes_ under the same [MIT License](
  https://github.com/opencv/cvat/blob/develop/LICENSE) that covers the project.
  Feel free to contact the maintainers if that's a concern.

### CODE REVIEW

File: cvat-core/src/api-implementation.ts [hunk 1]
```diff
@@ -33,39 +33,40 @@ import QualityConflict from './quality-conflict';
...
-    cvat.server.share.implementation = async (directory: string, searchPrefix?: string) => {
+    });
+    implementationMixin(cvat.server.share, async (directory: string, searchPrefix?: string) => {
         const result = await serverProxy.server.share(directory, searchPrefix);
         return result.map((item) => ({ ...omit(item, 'mime_type'), mimeType: item.mime_type }));
-    };
-
-    cvat.server.formats.implementation = async () => {
+    });
+    implementationMixin(cvat.server.formats, async () => {
```
Comment (klakhov) [inline cvat-core/src/api-implementation.ts:61]: Please, add newlines inbetween the implementations. Or maybe just between the groups cvat.`something`

File: cvat-core/src/download.worker.ts
Comment (klakhov) [inline cvat-core/src/download.worker.ts:1]: Header update?

File: cvat-core/src/index.ts [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+    server: {
+        about: typeof serverProxy.server.about;
+        share: (dir: string) => Promise<{
+            mimeType: string;
+            name: string;
+            type: enums.ShareFileType;
+        }[]>;
+        formats: () => Promise<AnnotationFormats>;
+        userAgreements: typeof serverProxy.server.userAgreements,
+        register: any; // TODO: add types later
```
Comment (klakhov) [inline cvat-core/src/index.ts:54]: Will it be added in this patch or wait for next one?

File: cvat-ui/src/components/import-dataset/import-dataset-modal.tsx [hunk 3]
```diff
@@ -234,12 +234,15 @@ function ImportDatasetModal(props: StateToProps): JSX.Element {
...
-            ) as any);
+            dispatch(
+                importDatasetAsync(
+                    instance,
+                    uploadParams.selectedFormat as string,
+                    uploadParams.useDefaultSettings,
+                    uploadParams.sourceStorage,
+                    uploadParams.file || uploadParams.fileName as string,
+                    uploadParams.convMaskToPoly,
+                ) as any);
```
Comment (klakhov) [inline cvat-ui/src/components/import-dataset/import-dataset-modal.tsx:245]: I dont really think we need this `any`

File: cvat-core/src/index.ts (hunk 2 above)
Comment (bsekachev) [reply cvat-core/src/index.ts:54]: I was planning to add these types gradually. It is too big task for one PR

File: cvat-core/src/api-implementation.ts (hunk 1 above)
Comment (bsekachev) [reply cvat-core/src/api-implementation.ts:61]: Right now there are newlines exactly between groups, like between cvat.server and cvat.assets, etc. 
If to add these spaces between each row, we will lose some kind of visual difference


We can't add 2 new blank lines, because eslint complaints about it

File: cvat-data/src/ts/3rdparty/README.md [hunk 4]
```diff
@@ -1,7 +1,7 @@
 ## 3rdparty components
 
 These files are from the [Broadway.js](https://github.com/mbebenita/Broadway) repository:
-- Decoder.js
+- Decoder.cjs
```
Comment (SpecLad) [inline cvat-data/src/ts/3rdparty/README.md:4]: There is no `Decoder.cjs`.

File: cvat-data/src/ts/3rdparty_patch.diff
Comment (SpecLad) [inline cvat-data/src/ts/3rdparty_patch.diff:1]: Why did you delete this?

File: cvat-canvas/webpack.config.cjs
Comment (SpecLad) [inline cvat-canvas/webpack.config.cjs:1]: Does webpack not support ESM configs?

File: cvat-data/src/ts/3rdparty_patch.diff
Comment (bsekachev) [reply cvat-data/src/ts/3rdparty_patch.diff:1]: Because we did not use it

File: cvat-canvas/webpack.config.cjs
Comment (bsekachev) [reply cvat-canvas/webpack.config.cjs:1]: I didn't really understand your question

File: cvat-core/src/api-implementation.ts (hunk 1 above)
Comment (klakhov) [reply cvat-core/src/api-implementation.ts:61]: Ok, I see

File: cvat-data/src/ts/3rdparty_patch.diff
Comment (bsekachev) [reply cvat-data/src/ts/3rdparty_patch.diff:1]: - Initially we used this patch file to patch the library during build
- As I discovered, patched version was already commited long time ago and we don't use the file anymore
- Finally I do not really see a reason to patch this file, their repository is dead, we don't receive any updates. License allows us to modify it, so..

File: cvat-canvas/webpack.config.cjs
Comment (bsekachev) [reply cvat-canvas/webpack.config.cjs:1]: If you mean webpack.config file is a common js module, so, I decided to leave configurations as they are. 
They write than webpack should support ESM configurations in our version. 

But it lead to issues with eslint/jest configurations when I tried to convert them to ESM.

Comment (SpecLad) [reply cvat-canvas/webpack.config.cjs:1]: Okay.
